File: php/tests/frame.test.php
<?php
// Frame classification tests. Verdicts mirror the shared corpus
// (spec/corpus/framing.json). Byte-for-byte agreement over that corpus is
// checked by bin/interop_checks.php.
use Bonemesh\Frame;

test('accepts a simple object', function () {
    $r = Frame::classify('{"t":"bmx1","v":3}' . "\n", Frame::HANDSHAKE_CAP);
    assertTrue(!isset($r['reason']));
    assertEq('bmx1', $r['obj']['t']);
});

test('verdicts match the corpus shapes', function () {
    $cases = [
        ['no-newline', '{"a":1}', Frame::TRANSPORT_CAP, 'no-newline'],
        ['empty', "\n", Frame::TRANSPORT_CAP, 'empty'],
        ['invalid-json garbage', "not json\n", Frame::TRANSPORT_CAP, 'invalid-json'],
        ['invalid-json interior-newline', '{"a":' . "\n" . '1}' . "\n", Frame::TRANSPORT_CAP, 'invalid-json'],
        ['trailing-data', '{"a":1} X' . "\n", Frame::TRANSPORT_CAP, 'trailing-data'],
        ['not-an-object', '[1,2,3]' . "\n", Frame::TRANSPORT_CAP, 'not-an-object'],
        ['oversize', '{"a":1}' . "\n", 4, 'oversize'],
    ];
    foreach ($cases as [$name, $raw, $cap, $want]) {
        assertEq($want, Frame::classify($raw, $cap)['reason'] ?? '(accept)', $name);
    }
});

test('rejects invalid utf-8', function () {
    $raw = '{"a":"' . chr(0xff) . chr(0xfe) . '"}' . "\n";
    assertEq('invalid-utf8', Frame::classify($raw, Frame::HANDSHAKE_CAP)['reason']);
});

test('nested braces inside strings scan correctly', function () {
    $r = Frame::classify('{"a":{"b":"}]"},"c":[1,{"d":2}]}' . "\n", Frame::TRANSPORT_CAP);
    assertTrue(!isset($r['reason']));
    assertEq([1, ['d' => 2]], $r['obj']['c']);
});

test('encode round-trips through classify', function () {
    $enc = Frame::encode(['t' => 'bmx1', 'mesh' => 'acme']);
    assertEq("\n", substr($enc, -1));
    assertEq('acme', Frame::classify($enc, Frame::HANDSHAKE_CAP)['obj']['mesh']);
});

File: rust/tests/frame_test.rs
// Frame classification tests. Cases mirror the shared corpus
// (spec/corpus/framing.json); byte-for-byte agreement with the Java, Go, and
// Elixir classifiers over that corpus is checked by interop/check-framing-rust.sh.

use bonemesh::frame::{classify, Verdict, TRANSPORT_CAP};

fn reason(raw: &[u8]) -> Option<&'static str> {
    match classify(raw, TRANSPORT_CAP) {
        Verdict::Accept(_) => None,
        Verdict::Reject(r) => Some(r),
    }
}

#[test]
fn simple_object_accepts() {
    assert_eq!(reason(b"{\"a\":1}\n"), None);
}
#[test]
fn missing_trailing_newline() {
    assert_eq!(reason(b"{\"a\":1}"), Some("no-newline"));
}
#[test]
fn empty_line() {
    assert_eq!(reason(b"\n"), Some("empty"));
}
#[test]
fn not_json() {
    assert_eq!(reason(b"not json\n"), Some("invalid-json"));
}
#[test]
fn json_array_is_not_an_object() {
    assert_eq!(reason(b"[1,2,3]\n"), Some("not-an-object"));
}
#[test]
fn interior_newline_splits_frame() {
    assert_eq!(reason(b"{\"a\":\n1}\n"), Some("invalid-json"));
}
#[test]
fn invalid_utf8() {
    assert_eq!(reason(&[b'{', b'"', b'a', b'"', b':', b'"', 0xff, b'"', b'}', b'\n']), Some("invalid-utf8"));
}
#[test]
fn trailing_garbage_after_object() {
    assert_eq!(reason(b"{\"a\":1} X\n"), Some("trailing-data"));
}
#[test]
fn frame_exactly_at_cap_accepts() {
    assert_eq!(reason(&line_of_length(TRANSPORT_CAP)), None);
}
#[test]
fn frame_one_byte_over_cap_is_oversize() {
    let raw = line_of_length(TRANSPORT_CAP + 1);
    assert_eq!(classify_reason_at_cap(&raw), Some("oversize"));
}

fn classify_reason_at_cap(raw: &[u8]) -> Option<&'static str> {
    match classify(raw, TRANSPORT_CAP) {
        Verdict::Accept(_) => None,
        Verdict::Reject(r) => Some(r),
    }
}

fn line_of_length(n: usize) -> Vec<u8> {
    let overhead = "{\"p\":\"\"}".len() + 1;
    let fill = n - overhead;
    let mut s = String::from("{\"p\":\"");
    s.push_str(&"A".repeat(fill));
    s.push_str("\"}\n");
    s.into_bytes()
}
